File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    answer = []
    total = {} # {장르 : 총 재생 횟수}
    genre = {} # {장르 : [(플레이 횟수, 고유번호)]}
    
    for i in range(len(genres)):
        if genres[i] in total:
            total[genres[i]] += plays[i]
        else:
            total[genres[i]] = plays[i]
            
        if genres[i] in genre:
            genre[genres[i]] += [(plays[i], i)]
        else:
            genre[genres[i]] = [(plays[i], i)]
            
    # 재생 횟수 내림차순으로 장르 정렬
    sorted_genre = sorted(total.items(), key=lambda x : x[1], reverse = True)
    # 재생 횟수 기준 내림차순, 인덱스 기준 오름차순
    for gen, totalPlay in sorted_genre:
        genre[gen] = sorted(genre[gen], key=lambda x: (-x[0], x[1]))
        answer += [idx for play, idx in genre[gen][:2]]
    return answer

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;
class Solution {
    public int[] solution(String[] genres, int[] plays) {
        int[] answer = {};
        List<Integer> arr = new ArrayList<>();
        Map<String, Integer> genre = new HashMap<>(); // 각 장르별 총 재생 횟수
        Map<String, Map<Integer, Integer>> music = new HashMap<>(); // 각 장르에 속하는 노래 및 재생 횟수
        for (int i = 0; i < plays.length; i++) {
            if (!genre.containsKey(genres[i])) {
                Map<Integer, Integer> map = new HashMap<>();
                map.put(i, plays[i]);
                music.put(genres[i], map);
                genre.put(genres[i], plays[i]);
            } else {
                music.get(genres[i]).put(i, plays[i]);
                genre.put(genres[i], genre.get(genres[i]) + plays[i]);
            }
        }
        
        List<String> keySet = new ArrayList(genre.keySet());
        Collections.sort(keySet, (k1, k2) -> genre.get(k2) - genre.get(k1));
        for (String key : keySet) {
            Map<Integer, Integer> map = music.get(key);
            List<Integer> genre_key = new ArrayList(map.keySet());
            
            Collections.sort(genre_key, (s1, s2) -> map.get(s2) - map.get(s1));
            
            arr.add(genre_key.get(0));
            if (genre_key.size() > 1) {
                arr.add(genre_key.get(1));
            }
        }
        
        answer = new int[arr.size()];
        for (int i = 0; i < arr.size(); i++) {
            answer[i] = arr.get(i);
        }
        
        return answer;
    }
}